### PR TITLE
[package] Make exporters write to buffer in fbcode

### PR DESCRIPTION
**Pull request by SplitInfinity**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54303 [package] Make exporters write to buffer in fbcode**

**Summary**
Creating temporary files can cause problem in fbcode. This commit
updates the packaging tests so that exporters write to a memory
buffer when tests run in fbcode.

**Test Plan**
Continuous integration.

Differential Revision: [D27180839](https://our.internmc.facebook.com/intern/diff/D27180839)

### CODE REVIEW

File: torch/package/package_exporter.py [hunk 1]
```diff
@@ -615,6 +615,11 @@ def close(self):
         if self.buffer:
             self.buffer.flush()
 
+        # If self.buffer is a BytesIO, rewind it in anticipation
+        # of a read in the near future.
+        if isinstance(self.buffer, io.BytesIO):
+            self.buffer.seek(0)
```
Comment (SplitInfinity) [inline torch/package/package_exporter.py:621]: I don't know if this is really that great of an idea, but it does help obviate the need to add `buffer.seek(0)` all over the place in unit tests.

Comment (suo) [reply torch/package/package_exporter.py:621]: Probably not a good idea.
- Convention of Python standard library is not to do this (in particular `pickle` does not do this)
- We shouldn't assume ownership over the io stream.

File: test/package/common.py [hunk 2]
```diff
@@ -11,6 +12,10 @@ def __init__(self, *args, **kwargs):
         self._temporary_files = []
 
     def temp(self):
+        # Use a buffer instead of a temporary file in fbcode.
```
Comment (suo) [inline test/package/common.py:15]: I'm not sure if we want to have the behavior be different between fbcode/external. The benefit is that we get both paths tested well, the disadvantage is that all test code needs to handle both cases, and it makes the test logic slightly more confusing (not obvious that behavior is different when debugging an individual test).

What about: use `BytesIO` everywhere as convention, but have the tests in `test_save_load`  explicitly test the file-based path and disable those tests in sandcastle? It's more work, but I think it's good to have the tests be very explicit. Could be convinced either way though.

**Review by suo — CHANGES_REQUESTED** [review]

File: test/package/common.py (hunk 2 above)
Comment (SplitInfinity) [reply test/package/common.py]: I made this change and pushed again. The only thing I don't like about it is that people writing tests have to remember to `buffer.seek(0)` all the time.

What about introducing a context manager that automatically rewinds the test buffer that its `__enter__` returns:
```
with PackageTestBuffer() as buffer:
    # use buffer in exporter

# buffer is automatically rewound
# use buffer in importer
```

It is not intuitive that buffer can be used outside of the with statement but Python doesn't have scope anyway...

Comment (suo) [reply test/package/common.py]: Dunno if that's much better, you have to remember to do something special either way. At least `io` is a standard thing.

**Review by suo — APPROVED** [review]

**Pull request closed without merging**